firmware/ventilator-controller-stm32/Core/Test/Pufferfish/Util/vector.txt

```diff
@@ -0,0 +1,22 @@
+SCENARIO( " Vectors are resized")
```

Comment:
Since this file tests the `Vector` class, please use structure the SCENARIOs/GIVENs/WHENs/THENs with the standard patterns we are using for non-mutating methods and mutating methods ([documented on Notion](https://www.notion.so/pufferfish/Testing-13ce58d72a154b6883fbd57b71d37f76#51087c7b9af449d79163521b431a2e28) as patterns 1 and 2).

Reply:
Tests should have coverage of all versions of all methods in the `Vector` class. In general, tests should cover when things (inputs and, if possible, internal state) are out-of-bounds (i.e. index > size), near bounds (i.e. index == size, index == size - 1, and index == size + 1), and within bounds (i.e. index < size - 1). Tests should have coverage of the behaviors of all methods when the vector has size 0, as well as when the vector is partially filled, as well as when the vector is completely full.

Tests need to be especially rigorous with the `erase` and `copy_from` methods, since that is where the logic is the most complicated and where hidden off-by-one errors, bounds-errors, and silent illegal accesses in the implementations are most likely to have occurred when I implemented those methods.

We also need tests confirming that the vector's state is correct when we directly modify the internal buffer using `Element *buffer()` and `resize` instead of using `push_back` and `erase`; this includes testing that the internal data should be valid after operations using those methods, and that the other methods (e.g. `push_back` and `erase`) work correctly after `Element *buffer()` and `resize` are used.

Reply:
For example, we need tests covering the behavior and state of the vector after `operator[]` is called with an out-of-bounds index (e.g. vector has capacity 20, is currently holding 10 elements, and then the `[]` operator is called for index 10, and for index 15).

Reply:
@raavilagoo For `Vector` methods which have something related to the internal state of a `Vector` object (which I believe is all methods in the `Vector` class), please use test [pattern 2](https://www.notion.so/pufferfish/Testing-13ce58d72a154b6883fbd57b71d37f76#ffd035643205441ebb0ea3cdef473d8d) from the Notion document. In other words, the `GIVEN`s should be objects, not methods.

Reply:
The approach to writing these tests should be to assume that each method of the object is or will become complicated and buggy, and tests should cover enough of the behavior of those methods that we can detect already-existing bugs and future regressions.

Reply:
We still need tests confirming that the vector's state is correct when we directly modify the internal buffer using `Element *buffer()` instead of using `push_back` and `erase`. Modifying data by using the `buffer` method should not modify the size of the `Vector` object, only the values inside the `Vector` at the indices which were touched using the `buffer` method.